src/box/txn_limbo.c

```diff
@@ -40,6 +40,67 @@
 
 struct txn_limbo txn_limbo;
 
+static int
+txn_limbo_write_synchro(struct txn_limbo *limbo, uint16_t type, int64_t lsn,
+			uint64_t term, struct vclock *vclock);
+
+static void
+txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn);
+
+/**
+ * Write a confirmation entry to the WAL. After it's written all the
+ * transactions waiting for confirmation may be finished.
+ */
+static int
+txn_limbo_write_confirm(struct txn_limbo *limbo, int64_t lsn)
+{
+	assert(lsn > limbo->confirmed_lsn);
+	assert(!limbo->is_in_rollback);
+	return txn_limbo_write_synchro(limbo, IPROTO_RAFT_CONFIRM, lsn, 0,
+				       NULL);
+}
+
+static int
+txn_limbo_worker_bump_confirmed_lsn(struct txn_limbo *limbo)
+{
+	assert(limbo->is_in_rollback ||
+	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
+	while (txn_limbo_is_owned_by_current_instance(limbo) &&
+	       limbo->volatile_confirmed_lsn > limbo->confirmed_lsn) {
+		if (limbo->is_in_rollback)
+			return -1;
+		/* It can get bumped again while we are writing. */
+		int64_t volatile_confirmed_lsn = limbo->volatile_confirmed_lsn;
+		if (txn_limbo_write_confirm(limbo,
+					    volatile_confirmed_lsn) != 0) {
+			diag_log();
+			return -1;
+		}
+		ERROR_INJECT_YIELD(ERRINJ_TXN_LIMBO_WORKER_DELAY);
+		txn_limbo_read_confirm(limbo, volatile_confirmed_lsn);
+	}
+	assert(limbo->is_in_rollback ||
+	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
+	return 0;
+}
+
+static int
+txn_limbo_worker_f(va_list args)
+{
+	(void)args;
+	struct txn_limbo *limbo = fiber()->f_arg;
+	assert(limbo == &txn_limbo);
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+		ERROR_INJECT_YIELD(ERRINJ_TXN_LIMBO_WORKER_DELAY);
+		if (txn_limbo_worker_bump_confirmed_lsn(limbo) != 0)
+			fiber_sleep(1);
```

Comment:
Lets make it `0.01` when `#if TEST_BUILD`? (It is set in `trivial/config.h` for all non-packaged builds, including Release build used in CI.)

Reply:
I don't like the idea of having the system behave differently under debug and release build types.

Reply:
That will work both in Release and Debug. `TEST_BUILD` doesn't depend on the build type. It is set for local and CI builds allowing to make them faster.

Reply:
Sorry, my bad, I misread your comment, thought you were just pushing for this once again https://github.com/tarantool/tarantool/pull/10626#discussion_r1781690963.

Applied your suggestion.